### PR TITLE
Suit sensor and crew monitoring

### PR DESCRIPTION
**Pull request by Macoron**
## About the PR 
Add basic crew monitoring console and suit sensors.

It's a bit experiential, because suit sensor use device network to broadcast their state to crew monitoring consoles. It works like this:
1. Each two seconds sensor updates its state
2. State is packed into device network message and broadcasted over wireless net
3. Monitoring console receive message and save it in an internal buffer
4. If console hasn't received an state update from sensor for more than 10 seconds, it presumes that sensor was disabled or lost connection and remove it from buffer

As a side effect of last part: if you quickly change character jumpsuits,  crew monitoring console will display duplicates (though they will disappear quickly).

**Screenshots**

image

image

**Changelog**

:cl:
- add: Added suit sensors and crew monitoring console in code. They are not mapped, but you can print circuit board.
- add: Prisoners suit has locked suit sensor that always broadcasting their position.

### CODE REVIEW

**Comment by moonheart08** [comment]
Probably needs adjusted for [5422] should that be merged first

File: Content.Server/Medical/SuitSensors/SuitSensorSystem.cs
Comment (juliangiebel) [inline Content.Server/Medical/SuitSensors/SuitSensorSystem.cs:265]: It's called packet so you should probably name that method PacketToSuitSensor 😄

File: Resources/Prototypes/Entities/Clothing/Uniforms/base.yml
Comment (juliangiebel) [inline Resources/Prototypes/Entities/Clothing/Uniforms/base.yml:10]: Did you try out the range in game?

**Review by juliangiebel — APPROVED** [review]
Looks good to me 👍 
Usage off device networking is good 👌

File: Resources/Prototypes/Entities/Clothing/Uniforms/base.yml
Comment (Macoron) [reply Resources/Prototypes/Entities/Clothing/Uniforms/base.yml:10]: Do you mean out of range? Yes, worked fine

**Review by PaulRitter — CHANGES_REQUESTED** [review]
checking to see if my pending comment got eaten by the github crash

File: Content.Client/Medical/CrewMonitoring/CrewMonitoringTableEntry.xaml.cs
Comment (PaulRitter) [inline Content.Client/Medical/CrewMonitoring/CrewMonitoringTableEntry.xaml.cs:16]: i dont see this control used anywhere?

File: Content.Server/Medical/SuitSensors/SuitSensorSystem.cs
Comment (PaulRitter) [inline Content.Server/Medical/SuitSensors/SuitSensorSystem.cs:80]: this should be done on componentinit, not mapinit imo

File: Content.Server/Medical/SuitSensors/SuitSensorComponent.cs
Comment (PaulRitter) [inline Content.Server/Medical/SuitSensors/SuitSensorComponent.cs:53]: pretty sure you can just get the currect wearer by getting the clothings parent (Transform.ParentUid)

File: Content.Server/Medical/SuitSensors/SuitSensorSystem.cs
Comment (PaulRitter) [inline Content.Server/Medical/SuitSensors/SuitSensorSystem.cs:140]:
```suggestion
                CreateVerb(uid, component, args.User.Uid, SuitSensorMode.SensorVitals),
```

File: Content.Server/Medical/SuitSensors/SuitSensorSystem.cs
Comment (PaulRitter) [inline Content.Server/Medical/SuitSensors/SuitSensorSystem.cs:242]:
```suggestion
            // finally, form suit sensor status
            var status = new SuitSensorStatus(userName, userJob);
            switch (sensor.Mode)
            {
                case SuitSensorMode.SensorCords:
                    status.Coordinates = transform.MapPosition;
                case SuitSensorMode.SensorVitals:
                    // get mob total damage
                    if (EntityManager.TryGetComponent(sensor.User.Value, out DamageableComponent? damageable))
                        status.TotalDamage = damageable.TotalDamage.Int();
                case SuitSensorMode.SensorBinary:
                    // get health mob state
                    if (EntityManager.TryGetComponent(sensor.User.Value, out MobStateComponent? mobState))
                        status.IsAlive = mobState.IsAlive();
                    break;
            }
```
webedited together, check formatting after you commit this one or just code it yourself

**Review by PaulRitter — CHANGES_REQUESTED** [review]
it did

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/Medical/SuitSensors/SuitSensorSystem.cs
Comment (Macoron) [reply Content.Server/Medical/SuitSensors/SuitSensorSystem.cs:80]: what the difference?

File: Content.Server/Medical/SuitSensors/SuitSensorComponent.cs
Comment (Macoron) [reply Content.Server/Medical/SuitSensors/SuitSensorComponent.cs:53]: I don't think its a good idea. Sounds like a dirty hack.
Equipped and Unequipped events are much more robust solution. It also allows to check that sensor was actually worn in a valid inventory slot.

File: Content.Server/Medical/SuitSensors/SuitSensorSystem.cs
Comment (Macoron) [reply Content.Server/Medical/SuitSensors/SuitSensorSystem.cs:242]: I've checked in IDE. Code above isn't valid C# code. You can't "fallthrough" case statements in C#. Here is an example which wouldn't compily:
```
private enum X { A, B }

private void Boo()
{
    int i = 0;
    var e = X.A;

    switch (e)
    {
        case X.A:
            i++;
        case X.B:
            i--;
            break;
    }
}
```

Comment (PaulRitter) [reply Content.Server/Medical/SuitSensors/SuitSensorSystem.cs:242]: my bad, you need to write `goto case X;` to create a fallthrough. implicit fallthroughs are disabled in c#, for good reason i might add, so you need to explicitely define them.

that being said, i forgot to mention why i actually wanted this change: i dont like simply a mobstatecomp/damageablecomp not existing killing the entire state from being sent.

Comment (Macoron) [reply Content.Server/Medical/SuitSensors/SuitSensorSystem.cs]: I've added default values to current code, if components don't exist.

**Comment by Macoron** [comment]
Addressed reviews and fixed merge conflict.

> Probably needs adjusted for #5422 should that be merged first

Tbh, I'm not sure, how to adjust this for multistation setup. Right now any suit sensor will be shown in crew monitoring console,  if they are in 500 tiles radius. The mob name and job are directly taken from IdCardSlot. Should I filter mobs somehow by grid id? Or maybe it would be better to use different wireless network channels?

If someone has an idea what should be multistation crew monitoring behavior, I can create an issue, after this PR will be merged.

**Comment by moonheart08** [comment]
> Addressed reviews and fixed merge conflict.
> 
> > Probably needs adjusted for #5422 should that be merged first
> 
> Tbh, I'm not sure, how to adjust this for multistation setup. Right now any suit sensor will be shown in crew monitoring console, if they are in 500 tiles radius. The mob name and job are directly taken from IdCardSlot. Should I filter mobs somehow by grid id? Or maybe it would be better to use different wireless network channels?
> 
> If someone has an idea what should be multistation crew monitoring behavior, I can create an issue, after this PR will be merged.

What could be done is simply including the StationId the sensor was originally registered on in the message, and only showing it in the console on the corresponding station. You can check what station a  grid is part of by grabbing the id from the grid's StationComponent

**Comment by Macoron** [comment]
> > Addressed reviews and fixed merge conflict.
> > > Probably needs adjusted for #5422 should that be merged first
> > 
> > 
> > Tbh, I'm not sure, how to adjust this for multistation setup. Right now any suit sensor will be shown in crew monitoring console, if they are in 500 tiles radius. The mob name and job are directly taken from IdCardSlot. Should I filter mobs somehow by grid id? Or maybe it would be better to use different wireless network channels?
> > If someone has an idea what should be multistation crew monitoring behavior, I can create an issue, after this PR will be merged.
> 
> What could be done is simply including the StationId the sensor was originally registered on in the message, and only showing it in the console on the corresponding station. You can check what station a grid is part of by grabbing the id from the grid's StationComponent

I've discussed this issue in Discord. This would probably solved by new encryption keys mechanics for different stations in a separate PR.

**Comment by PaulRitter** [comment]
> > > Addressed reviews and fixed merge conflict.
> > > > Probably needs adjusted for #5422 should that be merged first
> > > 
> > > 
> > > Tbh, I'm not sure, how to adjust this for multistation setup. Right now any suit sensor will be shown in crew monitoring console, if they are in 500 tiles radius. The mob name and job are directly taken from IdCardSlot. Should I filter mobs somehow by grid id? Or maybe it would be better to use different wireless network channels?
> > > If someone has an idea what should be multistation crew monitoring behavior, I can create an issue, after this PR will be merged.
> > 
> > 
> > What could be done is simply including the StationId the sensor was originally registered on in the message, and only showing it in the console on the corresponding station. You can check what station a grid is part of by grabbing the id from the grid's StationComponent
> 
> I've discussed this issue in Discord. This would probably solved by new encryption keys mechanics for different stations in a separate PR.

if you want to implement that, feel free to. I'd be happy with you doing what moony said, considering the encryption keys out of scope. Your call.

**Comment by Macoron** [comment]
> > > > Addressed reviews and fixed merge conflict.
> > > > > Probably needs adjusted for #5422 should that be merged first
> > > > 
> > > > 
> > > > Tbh, I'm not sure, how to adjust this for multistation setup. Right now any suit sensor will be shown in crew monitoring console, if they are in 500 tiles radius. The mob name and job are directly taken from IdCardSlot. Should I filter mobs somehow by grid id? Or maybe it would be better to use different wireless network channels?
> > > > If someone has an idea what should be multistation crew monitoring behavior, I can create an issue, after this PR will be merged.
> > > 
> > > 
> > > What could be done is simply including the StationId the sensor was originally registered on in the message, and only showing it in the console on the corresponding station. You can check what station a grid is part of by grabbing the id from the grid's StationComponent
> > 
> > 
> > I've discussed this issue in Discord. This would probably solved by new encryption keys mechanics for different stations in a separate PR.
> 
> if you want to implement that, feel free to. I'd be happy with you doing what moony said, considering the encryption keys out of scope. Your call.

Tbh, I don't see any issue with current implementation. Not a big deal if players can see other players from another station in 500 tiles radius. Sounds more like a feature to me.

So I would like to leave this PR as is and address it in other PR.

**Comment by mirrorcult** [comment]
leaving this as-is and encryption keys for another PR sounds fine to me

**Comment by moonheart08** [comment]
> leaving this as-is and encryption keys for another PR sounds fine to me

Sounds fine to me.

File: Content.Server/Medical/SuitSensors/SuitSensorSystem.cs
Comment (PaulRitter) [reply Content.Server/Medical/SuitSensors/SuitSensorSystem.cs:90]: both work, but i'd still rather have it on compinit since its not maprelated.

compinit -> when the component gets intialized on entitycreation
mapinit -> when the map the entity is on gets intialized

Comment (Macoron) [reply Content.Server/Medical/SuitSensors/SuitSensorSystem.cs:90]: > both work, but i'd still rather have it on compinit since its not maprelated.

But why? Right now if you save and load map - suit sensor will save its mode. But if you move it to init, it will be rewriting its state with each reload.

**Comment by Zumorica** [comment]
Conflicts

File: Content.Server/Medical/SuitSensors/SuitSensorSystem.cs
Comment (PaulRitter) [reply Content.Server/Medical/SuitSensors/SuitSensorSystem.cs:90]: afaict mapinit should still run, even if you load a map

Comment (PaulRitter) [reply Content.Server/Medical/SuitSensors/SuitSensorSystem.cs:90]: the idea is to have this run everytime a component is intialized, not everytime the map the components entity is on get initialized, since the first makes more sense

**Comment by Macoron** [comment]
Oh, thanks Sloth. I was too busy on my work, didn't have time to fix merge conflict by myself. Thank you for your help.

**Pull request merged**
